File: app/src/main/java/de/xikolo/controllers/course_items/RichTextFragment.java
package de.xikolo.controllers.course_items;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.yatatsu.autobundle.AutoBundleField;

import butterknife.BindView;
import butterknife.OnClick;
import de.xikolo.R;
import de.xikolo.controllers.base.LoadingStatePresenterFragment;
import de.xikolo.controllers.webview.WebViewActivityAutoBundle;
import de.xikolo.models.Item;
import de.xikolo.models.RichText;
import de.xikolo.presenters.base.PresenterFactory;
import de.xikolo.presenters.course_items.RichTextPresenter;
import de.xikolo.presenters.course_items.RichTextPresenterFactory;
import de.xikolo.presenters.course_items.RichTextView;
import de.xikolo.utils.MarkdownUtil;

public class RichTextFragment extends LoadingStatePresenterFragment<RichTextPresenter, RichTextView> implements RichTextView {

    public static final String TAG = RichTextFragment.class.getSimpleName();

    @AutoBundleField String courseId;
    @AutoBundleField String sectionId;
    @AutoBundleField String itemId;

    @BindView(R.id.title) TextView title;
    @BindView(R.id.text) TextView text;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public int getLayoutResource() {
        return R.layout.content_richtext;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void setupView(Item item, RichText richText) {
        title.setText(item.title);
        MarkdownUtil.formatAndSet(richText.text, text);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int itemId = item.getItemId();
        switch (itemId) {
            case R.id.action_refresh:
                presenter.onRefresh();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @SuppressWarnings("unused")
    @OnClick(R.id.fallback_button)
    public void fallbackButtonClicked() {
        presenter.fallbackButtonClicked();
    }

    public void openAsWebView(String title, String url, boolean inAppLinksEnabled, boolean externalLinksEnabled) {
        Intent intent = WebViewActivityAutoBundle.builder(title, url)
                .inAppLinksEnabled(inAppLinksEnabled)
                .externalLinksEnabled(externalLinksEnabled)
                .build(getActivity());
        startActivity(intent);
    }

    @NonNull
    @Override
    protected PresenterFactory<RichTextPresenter> getPresenterFactory() {
        return new RichTextPresenterFactory(courseId, sectionId, itemId);
    }

}

File: app/src/main/java/de/xikolo/controllers/section/RichTextFragment.kt
package de.xikolo.controllers.section

import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import butterknife.BindView
import com.yatatsu.autobundle.AutoBundleField
import de.xikolo.R
import de.xikolo.config.Config
import de.xikolo.controllers.base.NetworkStateFragment
import de.xikolo.controllers.webview.WebViewActivityAutoBundle
import de.xikolo.extensions.observe
import de.xikolo.models.Item
import de.xikolo.utils.LanalyticsUtil
import de.xikolo.utils.MarkdownUtil
import de.xikolo.viewmodels.section.RichTextViewModel

class RichTextFragment : NetworkStateFragment<RichTextViewModel>() {

    companion object {
        val TAG: String = RichTextFragment::class.java.simpleName
    }

    @AutoBundleField
    lateinit var courseId: String

    @AutoBundleField
    lateinit var sectionId: String

    @AutoBundleField
    lateinit var itemId: String

    @BindView(R.id.title)
    lateinit var title: TextView

    @BindView(R.id.text)
    lateinit var text: TextView

    @BindView(R.id.fallback_button)
    lateinit var fallbackButton: Button

    override val layoutResource = R.layout.content_richtext

    private var item: Item? = null

    override fun createViewModel(): RichTextViewModel {
        return RichTextViewModel(itemId)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        fallbackButton.setOnClickListener {
            openAsWebView(
                item?.title ?: "",
                Config.HOST_URL + Config.COURSES + courseId + "/" + Config.ITEMS + itemId,
                false,
                false
            )

            LanalyticsUtil.trackRichTextFallback(itemId, courseId, sectionId)
        }

        viewModel.item
            .observe(this) { item ->
                this.item = item
                title.text = item.title

                MarkdownUtil.formatAndSet(viewModel.richText?.text, text)
                showContent()
            }
    }

    private fun openAsWebView(title: String, url: String, inAppLinksEnabled: Boolean, externalLinksEnabled: Boolean) {
        activity?.let {
            val intent = WebViewActivityAutoBundle.builder(title, url)
                .inAppLinksEnabled(inAppLinksEnabled)
                .externalLinksEnabled(externalLinksEnabled)
                .build(it)
            startActivity(intent)
        }
    }

}
